types/wordpress__block-editor/wordpress__block-editor-tests.tsx

```diff
@@ -534,3 +535,18 @@ select('core/block-editor').getAdjacentBlockClientId();
 select('core/block-editor').getAdjacentBlockClientId('foo');
 select('core/block-editor').getAdjacentBlockClientId('foo', -1);
 select('core/block-editor').getAdjacentBlockClientId('foo', 1);
+
+// $ExpectType Omit<Record<string, unknown>, "ref"> & Merged & Reserved
+be.useBlockProps();
+
+// $ExpectType Omit<{ foo: string; }, "ref"> & Merged & Reserved
```

Comment:
this is probably causing the `never` here. a note on this applies to the other `$ExpectType` assertions.

what I think we're trying to do here is assert the _named_ types but those are lost through TypeScript's _structural_ type system. also I'm not sure where we're getting `Merged` and `Reserved` since those aren't imported in the test file (unless I'm overlooking that).

you might start by importing those types or referncing `be.Merged` and `be.Reserved`. alternatively we could test for specific expectations, such as `$ExpectType {foo: string}`

we're not checking that the implementation of the types doesn't change; we're rather verifying that the types in our implementation get us to the end result here.

Reply:
But this would not explain why, with the commit I just reverted, things works fine.

Moreover, I just tried adding
```typescript
export * as BlockProps from './use-block-props';
```
into `components.index.d.ts` and
```typescript
type Merged = be.BlockProps.Merged;
type Reserved = be.BlockProps.Reserved;
```
just before the assertions, and the problem does not change. Following Hokkam's razor principle, I still think that if the issue we get is related to a transformation between `Omit` and `Pick`, probably this is the actual problem.

> we're not checking that the implementation of the types doesn't change; we're rather verifying that the types in our implementation get us to the end result here.

Good point, I added a couple of tests for this.

Reply:
Thanks for staying persistent on this. I'm confused on what the goal of the last changes were, for one, adding the `UseBlockProps` export to the main `index.d.ts` file vs. importing it directly in the test file; two, is it intermediate code you added there to write `_merged` and `_reserved` without using them?

Sorry for my lack of understanding here.

> I still think that if the issue we get is related to a transformation between Omit and Pick, probably this is the actual problem

It's very possible this is the case, and I've not wanted to suggest it can't be. My takeaway from that though isn't to change our types away from what we want (and what we know is accurate) but instead to relax the test cases where the test runner is the problem.

Reply:
>  I'm confused on what the goal of the last changes were, for one, adding the UseBlockProps export to the main index.d.ts file vs. importing it directly in the test file; two, is it intermediate code you added there to write _merged and _reserved without using them?

You rightly observed that we were only testing that `useBlockProps` was returning an intersection of custom-defined types, without actually testing them. Declaring `_merged` and `_reserved` tests whether the two objects are compatible with `Merged` and `Reserved` interfaces. However, because the test sees `@wordpress/block-editor` from _outside_, I need a way to make the two types `Merged` and `Reserved` reachable in some way without polluting the namespaces, and for this I used `UseBlockProps`. If you remember, I originally used mouthful type names in order to directly export them, this new approach is cleaner but it needs an _intermediate_ module/namespace to avoid future name collisions.

> but instead to relax the test cases where the test runner is the problem

Ok, but I don't get how we can relax the tests. Can you make an example that can checks the returned type and that passes the tests?

Reply:
> That's true, but I question the value of it. Reserved is generated from inside useBlockProps so I think the only time it can fail is when we make an intentional change and it's unlikely to be able to catch any accidental change. Merged is similar other than it's not entirely generated from the inside. I'd think what we care about constraining is the interface to useBlockProps, not an internal type that's helpful for it.

That's the point of almost all the tests in dts: you are only checking that you type implementation reflects what you expected, nothing more and nothing less. If you want, I can remove them (I realized I misunderstood you, in a previous discussion you just wanted to explain things a little more, not to point out about the missing validation of the internal types).

> We should be able to directly import Merged and Reserved from the test file without running it through index.d.ts though is my point. It's easy to avoid namespacing issues too by renaming the imports or using an alias for the module.

Very good point, TIL. I did not know it let me do that. :heart: 

> It's still checking the return type but verifying fewer properties of it. This is a compromise but might be worth it; the last thing I ever want any of us to do is give up on a better goal because the tools are deficient.

But using `Record` does not compile, for the exact same reason it does not compile with `Omit`: it does not performs a _real_ type comparison, it it just a comparison between string representation (which "obviously" differs).

Just let me know if you want me to remove the regression tests for `_merged` and `_reserved`, then I will remove the export of `UseBlockProps` following your suggestion. Let me know if you have any other idea to change the `ExpectType`.

Reply:
> But using Record does not compile, for the exact same reason it does not compile with Omit: it does not performs a real type comparison, it it just a comparison between string representation (which "obviously" differs).

Oy. I also wasn't thinking about this. Why? haha, and on a Friday! (or Saturday I guess depending on where you are).

How about we skirt the issue altogether. I think we can do this instead.

```ts
const testProps = be.useBlockProps();
// $ExpectType 'document'
testProps.role
// $ExpectType string
testProps.id
…
```

How do you feel about that? We can test as many or few of those properties as we deem important.

> Just let me know if you want me to remove the regression tests for _merged and _reserved, then I will remove the export of UseBlockProps following your suggestion

It's my own personal opinion that they add maintenance burden while failing to provide much safety or value. I am fine trusting your call; they don't weaken the type safety. 😄

Reply:
> How do you feel about that? We can test as many or few of those properties as we deem important.

I think that it is a pretty nice idea. I tried to use it in a sort of _mixed_ approach, in order to avoid a validation of `Merged` and `Reserved` instances (which would make things even worse in terms of maintainability). See what do you think about it.

> It's my own personal opinion that they add maintenance burden while failing to provide much safety or value. I am fine trusting your call; they don't weaken the type safety. smile

Ok, after thinking a bit about it, I decided to remove it, for one particular reason: it is not possible to express the `Exact<T>` type. This means that if someone will add a field to `Merged` or `Reserved` in the future, tests would pass anyway because the type is not _upper bounded_.

Reply:
I like how it turned out. Thanks for all your help! Thanks for enduring the exhausting back and forth 😄 

Let's get this in! :shipit: